nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/schema/inference/FieldTypeInference.java

```diff
@@ -62,36 +62,44 @@ public void addPossibleDataType(final DataType dataType) {
             final RecordSchema newSchema = ((RecordDataType) dataType).getChildSchema();
 
             final RecordSchema mergedSchema = DataTypeUtils.merge(singleDataTypeSchema, newSchema);
+            possibleDataTypes.remove(singleDataType);
             singleDataType = RecordFieldType.RECORD.getRecordDataType(mergedSchema);
+            possibleDataTypes.add(singleDataType);
             return;
         }
 
-        if (singleFieldType.isWiderThan(additionalFieldType)) {
-            // Assigned type is already wide enough to encompass the given type
-            return;
+        if (possibleDataTypes.isEmpty()) {
+            possibleDataTypes.add(singleDataType);
         }
 
-        if (additionalFieldType.isWiderThan(singleFieldType)) {
-            // The given type is wide enough to encompass the assigned type. So changed the assigned type to the given type.
-            singleDataType = dataType;
-            return;
-        }
+        boolean hasWiderNonString = possibleDataTypes.stream()
+                .map(DataType::getFieldType)
+                .filter(possibleDataType -> !possibleDataType.equals(RecordFieldType.STRING))
+                .filter(possibleDataType -> possibleDataType.isWiderThan(additionalFieldType))
+                .findAny()
+                .isPresent();
+
+        if (!hasWiderNonString) {
+            java.util.Iterator<DataType> possibleDataTypeIterator = possibleDataTypes.iterator();
```

Comment:
Should import `java.util.Iterator` rather than declaring the full package inline.

Reply:
Hmm, it must have been a glitch in the Matrix. :)